File: Algoritmos/Python/merge_sort.py
import sys
import time


# --- Funções do Merge Sort ---
def intercala(X, inicio, meio, fim):
    """
    Combina duas sub-listas ordenadas (X[inicio..meio] e X[meio+1..fim])
    em uma única lista ordenada.
    """
    aux = []
    inicio_vetor1 = inicio
    inicio_vetor2 = meio + 1

    # Intercala os elementos em ordem na lista auxiliar
    while inicio_vetor1 <= meio and inicio_vetor2 <= fim:
        if X[inicio_vetor1] <= X[inicio_vetor2]:
            aux.append(X[inicio_vetor1])
            inicio_vetor1 += 1
        else:
            aux.append(X[inicio_vetor2])
            inicio_vetor2 += 1

    # Adiciona os elementos restantes do primeiro subvetor (se houver)
    while inicio_vetor1 <= meio:
        aux.append(X[inicio_vetor1])
        inicio_vetor1 += 1

    # Adiciona os elementos restantes do segundo subvetor (se houver)
    while inicio_vetor2 <= fim:
        aux.append(X[inicio_vetor2])
        inicio_vetor2 += 1

    # Copia a lista auxiliar ordenada de volta para a lista original
    for i in range(len(aux)):
        X[inicio + i] = aux[i]

def merge_sort(X, inicio, fim):
    """
    Função principal que aplica o algoritmo Merge Sort recursivamente.
    """
    if inicio < fim:
        meio = (inicio + fim) // 2
        merge_sort(X, inicio, meio)
        merge_sort(X, meio + 1, fim)
        intercala(X, inicio, meio, fim)

def ler_vetor(nome_arquivo):
    
    with open(nome_arquivo, "r") as arquivo:
            vetor = [int(linha.strip()) for linha in arquivo]  # List comprehension
            return vetor

def main():
    
    nome_arquivo = sys.argv[1]

    vetor = ler_vetor(nome_arquivo)

    inicio = time.monotonic()  
    merge_sort(vetor, 0, len(vetor) - 1)
    fim = time.monotonic()
    tempo_execucao = fim - inicio  

    print(f"{tempo_execucao}")


if __name__ == "__main__":
    main()

File: Algoritmos/C++/merge_sort.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <filesystem> 
#include <vector>
#include <sstream>
#include <algorithm> 
#include <chrono>
#include <iomanip> 

void intercala(std::vector<int>& X, int inicio, int meio, int fim) {
    int i;
    int inicio_vetor1 = inicio;
    int inicio_vetor2 = meio + 1;
    
    std::vector<int> aux;
    aux.reserve(fim - inicio + 1);

    while (inicio_vetor1 <= meio && inicio_vetor2 <= fim) {
        if (X[inicio_vetor1] <= X[inicio_vetor2]) {
            aux.push_back(X[inicio_vetor1]);
            inicio_vetor1 = inicio_vetor1 + 1;
        } else {
            aux.push_back(X[inicio_vetor2]);
            inicio_vetor2 = inicio_vetor2 + 1;
        }
    }

    while (inicio_vetor1 <= meio) {
        aux.push_back(X[inicio_vetor1]);
        inicio_vetor1 = inicio_vetor1 + 1;
    }

    while (inicio_vetor2 <= fim) {
        aux.push_back(X[inicio_vetor2]);
        inicio_vetor2 = inicio_vetor2 + 1;
    }
    
    for (i = 0; i < aux.size(); i++) {
        X[inicio + i] = aux[i];
    }
    
}

void merge_sort(std::vector<int>& X, int inicio, int fim) {
    if (inicio < fim) {
        int meio = inicio + (fim - inicio) / 2; 
        merge_sort(X, inicio, meio);
        merge_sort(X, meio + 1, fim);
        intercala(X, inicio, meio, fim);
    }
}



std::vector<int> lerVetorDeArquivo(const std::string& nomeDoArquivo) {
    std::ifstream arquivo(nomeDoArquivo);

    std::vector<int> vetor;
    int numero;
    while (arquivo >> numero) {
        vetor.push_back(numero);
    }

    return vetor;
}

int main(int, char *argv[]) {
    
    std::string nomeDoArquivoParaAbrir = argv[1];
    std::vector<int> vetor = lerVetorDeArquivo(nomeDoArquivoParaAbrir);

    auto inicio = std::chrono::steady_clock::now(); 
    merge_sort(vetor, 0, vetor.size() - 1);
    auto fim = std::chrono::steady_clock::now();   
    
    std::chrono::duration<double> duracao = fim - inicio;

    std::cout << std::fixed << std::setprecision(9)
              << duracao.count()
              << std::endl;
        
    return 0; 
}